File: Remove K Digits.cpp
class Solution {
  public:
   string removeKdigits(string S, int k) {
        stack<char>st;
        for(int j=0; j<S.size(); j++){
            while(!st.empty() and st.top()>S[j] and k){
                st.pop();
                k--;
            }
            st.push(S[j]);
        }
        if(st.empty())
            return "0";
        while(k--)
            st.pop();
        string res;
        while(!st.empty()){
            res.push_back(st.top());
            st.pop();
        }
        reverse(res.begin(),res.end());
        int j=0;
        while(res[j]=='0')
            j++;
        if(j==res.size())
            return "0";
        return res.substr(j);
   }
    };

File: Remove K Digits.java
class Solution {
    public String removeKdigits(String S, int K) {

        if(S.length()==K){
            return "0";
        }

        Stack <Character> q=new Stack<>();
        q.push(S.charAt(0));
        for(int i=1;i<S.length();i++){
            while(!q.isEmpty() && q.peek()>S.charAt(i) && K !=0){
                q.pop();
                K--;
            }
            q.push(S.charAt(i));
        }

        Stack <Character> q2=new Stack<>();
        while(!q.isEmpty()){
            while(K!=0){
                q.pop();
                K--;
            }
            q2.push(q.pop());
        }
        while(q2.size()!=1 && q2.peek()=='0'){
            q2.pop();
        }

        StringBuilder sb=new StringBuilder("");
        

        while(!q2.isEmpty()){
            sb.append(q2.pop());
        }
        return sb.toString();

}

}
